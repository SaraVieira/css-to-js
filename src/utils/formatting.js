import prettier from "prettier/standalone";
import prettierCss from "prettier/parser-postcss";
import prettierBabel from "prettier/parser-babel";

/**
 * Takes some CSS code and formats it.
 * @param {string} cssString string of CSS code
 */
export function formatCss(cssString) {
  const prettierOutput = prettier.format(cssString, {
    parser: "css",
    plugins: [prettierCss],
  });

  return prettierOutput.trim();
}

/**
 * Takes a JS object written as a string and formats it.
 * @param {string} objString string that defines a JS object
 */
export function formatObject(objString) {
  // Write object in a piece of JS code so Prettier knows how to format it
  let codeString = `let temp = ${objString};`;

  codeString = prettier.format(codeString, {
    parser: "babel",
    plugins: [prettierBabel],
  });

  let groups = codeString.match(/let temp = (\{.*\});/s);
  if (!groups || groups.length < 2) {
    throw new Error("Something went wrong when parsing Prettier output");
  }

  return groups[1].trim();
}

/**
 * Takes some props written in JSX and formats them so they can easily be
 * applied to a React component.
 * @param {string} propString JSX props as a single string
 */
export function formatProps(propString) {
  // Write the props in a component so Prettier knows how to format it
  let componentString = `<Temp ${propString} />`;

  componentString = prettier.format(componentString, {
    parser: "babel",
    plugins: [prettierBabel],
  });

  // Return the Prettier output but without the component tag
  let groups = componentString.match(/<Temp(.*)\/>/s);
  if (!groups || groups.length < 2) {
    throw new Error("Something went wrong when parsing Prettier output");
  }

  return groups[1]
    .trim()
    .split("\n")
    .map((line) => line.trim()) // remove indentation on each line
    .join("\n");
}
