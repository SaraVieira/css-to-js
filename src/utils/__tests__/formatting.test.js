import { formatCss, formatObject, formatProps } from "../formatting";

describe("formatting", () => {
  describe("formatCss()", () => {
    test("formats a single rule", () => {
      let expected = "some-property: value;";
      expect(formatCss("some-property: value;")).toBe(expected);
      expect(formatCss("\t  some-property : \t value ")).toBe(expected);
    });

    test("formats more complex CSS", () => {
      expect(
        formatCss(
          `div{
            some-prop :   value;
            whatever: 42px;
          }
          

          .class{}; #id { prop:\tval}

          ;;
          `
        )
      ).toMatchInlineSnapshot(`
        "div {
          some-prop: value;
          whatever: 42px;
        }

        .class {
        }
        #id {
          prop: val;
        }"
      `);
    });
  });

  describe("formatObject()", () => {
    test("formats an object with a single key/value pair", () => {
      let expected = `{ someProp: someValue }`;
      expect(formatObject(`{ someProp: someValue }`)).toBe(expected);
      expect(formatObject(`\t {someProp  : someValue}`)).toBe(expected);
    });

    test("formats an object with multiple key/value pairs", () => {
      let expected = `{ x: 1, y: 2 }`;
      expect(formatObject(`{ x: 1, y: 2 }`)).toBe(expected);
      expect(formatObject(`{x:1,\n \ty: 2}  `)).toBe(expected);
    });

    test("formats a complex object", () => {
      expect(
        formatObject(
          `{ someProp: someValue,
            couldBe: {
              nested  : true, \t messedUp :true,
            }, anotherProp: 'string' || "string",
            ...{ spreadProp:  anyExpr ? what.ever:1337}
          }`
        )
      ).toMatchInlineSnapshot(`
        "{
          someProp: someValue,
          couldBe: {
            nested: true,
            messedUp: true,
          },
          anotherProp: \\"string\\" || \\"string\\",
          ...{ spreadProp: anyExpr ? what.ever : 1337 },
        }"
      `);
    });
  });

  describe("formatProps()", () => {
    test("formats a single prop with string value", () => {
      let expected = `someProp="some string"`;
      expect(formatProps(`someProp="some string"`)).toBe(expected);
      expect(formatProps(`\t  someProp= "some string" `)).toBe(expected);
    });

    test("formats a single prop with expression value", () => {
      let expected = "someProp={someExpr}";
      expect(formatProps("someProp={someExpr}")).toBe(expected);
      expect(formatProps("\t  someProp= {  someExpr } ")).toBe(expected);
    });

    test("formats multiple props", () => {
      expect(
        formatProps(
          `  someProp={someExpr} sneaky="beaky"
          anotherProp\t\t=  "whatever dude" 
          anotherOne= {{x: 2}}  `
        )
      ).toMatchInlineSnapshot(`
        "someProp={someExpr}
        sneaky=\\"beaky\\"
        anotherProp=\\"whatever dude\\"
        anotherOne={{ x: 2 }}"
      `);
    });
  });
});
