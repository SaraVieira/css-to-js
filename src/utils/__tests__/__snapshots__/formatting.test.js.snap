// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatting formatCss() formats more complex CSS 1`] = `
"div {
  some-prop: value;
  whatever: 42px;
}

.class {
}
#id {
  prop: val;
}"
`;

exports[`formatting formatObject() formats a complex object 1`] = `
"{
  someProp: someValue,
  couldBe: {
    nested: true,
    messedUp: true
  },
  anotherProp: \\"string\\" || \\"string\\",
  ...{ spreadProp: anyExpr ? what.ever : 1337 }
}"
`;

exports[`formatting formatProps() formats multiple props 1`] = `
"someProp={someExpr}
sneaky=\\"beaky\\"
anotherProp=\\"whatever dude\\"
anotherOne={{ x: 2 }}"
`;
